### PR TITLE
train-test-validation

### PR DESCRIPTION
**Pull request by pfistfl**

### CODE REVIEW

File: bookdown/02-basics-resampling.Rmd [hunk 1]
```diff
@@ -180,3 +180,169 @@ autoplot(rr, type = "roc")
...
+
+We are now going to use this instance to compare two versions of the `r ref("rpart")` learner with different hyperparameter settings, choose the better one and train this on the full data.
+
+```{r}
+resample(t, lrn("classif.rpart", cp = .1), rdesc)$score()
+resample(t, lrn("classif.rpart", cp = .5), rdesc)$score()
+
+# cp = 0.1 works better, train it and predict on validation rows.
+l = lrn("classif.rpart", cp = .1)$train(t)
+l$predict(t, row_ids = t$row_roles$validation)$score()
```
Comment (pfistfl) [inline bookdown/02-basics-resampling.Rmd:246]: use a measure that validates on validation data.

Comment (mllg) [reply bookdown/02-basics-resampling.Rmd:246]: You might need branch https://github.com/mlr-org/mlr3/tree/holdout :D

Comment (mllg) [reply bookdown/02-basics-resampling.Rmd:246]: Names are up for discussion

**Comment by pfistfl** [comment]
Currently relies on (https://github.com/mlr-org/mlr3/tree/holdout), will update once this is merged.

File: bookdown/02-basics-resampling.Rmd [hunk 2]
```diff
@@ -180,3 +180,169 @@ autoplot(rr, type = "roc")
...
+In order to tune and select hyperparameters, instead of using a **validation split**, we choose a **inner** resampling method that is used for this selection.
+This e.g. happens inside an `r ref("AutoTuner")` from `r mlr_pkg("mlr3tuning")`, where we provide an additional inner resampling method for tuning. If we now `r ref("resample()")` this `r ref("AutoTuner")`, we get the exact desired effect:
+The `r ref("AutoTuner")` internally tunes the  hyperparameters of our algorithm and `r ref("resample()")` provides us with an uniased estimate of this tuned algorithms performance.
+
+There are two approaches towards obtaining unbiased performance estimates from simple train-validation-test splits in **mlr3**.
+Either using a `r ref("AutoTuner")` together with `r ref("resample()")`, or using by keeping test data out using mlr3's `row_roles` mechanism.
+While the first approach is less flexible it is easy to understand and simple to apply. The second approach is slightly more complex, but also allows for a lot more flexibility and automation.
```
Comment (pat-s) [inline bookdown/02-basics-resampling.Rmd:204]:
```suggestion
While the first approach is less flexible it is easy to understand and simple to apply.
The second approach is slightly more complex, but also allows for a lot more flexibility and automation.
```

File: bookdown/02-basics-resampling.Rmd [hunk 3]
```diff
@@ -180,3 +180,169 @@ autoplot(rr, type = "roc")
...
+test = c(31:50, 81:100, 131:150)
+```
+
+We can now look at the `task`'s `row_roles`
+
+```{r}
+t$row_roles
+```
+
+We can see, that there is one set for `r names(t$row_roles)[1]`  and one for `r names(t$row_roles)[2]`. We can now set our `test` set to `validation` in order to keep it out from future computations.
```
Comment (pat-s) [inline bookdown/02-basics-resampling.Rmd:224]:
```suggestion
We can see, that there is one set for `r names(t$row_roles)[1]`  and one for `r names(t$row_roles)[2]`.
We can now set our `test` set to `validation` in order to keep it out from future computations.
```

File: bookdown/02-basics-resampling.Rmd [hunk 4]
```diff
@@ -180,3 +180,169 @@ autoplot(rr, type = "roc")
...
+
+```{r}
+t$set_row_role(test, "validation")
+```
+
+If we now create and instantiate a `holdout` resampling instance, we can see that
+those rows will not be used:
```
Comment (pat-s) [inline bookdown/02-basics-resampling.Rmd:231]:
```suggestion
If we now create and instantiate a `holdout` resampling instance, we can see that those rows will not be used:
```

File: bookdown/02-basics-resampling.Rmd [hunk 5]
```diff
@@ -180,3 +180,169 @@ autoplot(rr, type = "roc")
...
+
+# cp = 0.1 works better, train it and predict on validation rows.
+l = lrn("classif.rpart", cp = .1)$train(t)
+l$predict(t, row_ids = t$row_roles$validation)$score()
+```
+
+
+#### How can I get my pre-defined train-valid-test splits into mlr3?
+
+This example will make use of the `r ref("AutoTuner")`, described later on in the [tuning](#tuning) section. If you are not famililar with those concepts yet, it is advised to read up that section and come back here. Conceptually what happens here is similar to the example above, we split into fixed sets, select the better hyperparameter configuration and fit the final model.
```
Comment (pat-s) [inline bookdown/02-basics-resampling.Rmd:252]:
```suggestion
This example will make use of the `r ref("AutoTuner")`, described later on in the [tuning](#tuning) section.
If you are not famililar with those concepts yet, it is advised to read up that section and come back here.
Conceptually what happens here is similar to the example above, we split into fixed sets, select the better hyperparameter configuration and fit the final model.
```

File: bookdown/02-basics-resampling.Rmd [hunk 6]
```diff
@@ -180,3 +180,169 @@ autoplot(rr, type = "roc")
...
+Let us assume we have the following situation:
+
+```{r}
+t = tsk("iris")
+train = c(1:30, 51:80, 101:130)
+valid = c(25:40, 75:80, 125:140)
+test = c(41:50, 91:100, 141:150)
+```
+
+We now instantiate two holdout sets as done in the [custom resamplings section](#resamp-custom). In the outer loop, we use `train` and `valid` during training and the test set during testing.
```
Comment (pat-s) [inline bookdown/02-basics-resampling.Rmd:265]:
```suggestion
We now instantiate two holdout sets as done in the [custom resamplings section](#resamp-custom).
In the outer loop, we use `train` and `valid` during training and the test set during testing.
```

File: bookdown/02-basics-resampling.Rmd [hunk 7]
```diff
@@ -180,3 +180,169 @@ autoplot(rr, type = "roc")
...
+at = AutoTuner$new(
+  learner = learner,
+  resampling = inner,
+  measure = msr("classif.ce"),
+  search_space = tune_ps,
+  terminator = terminator,
+  tuner = tuner
+)
+```
+>> Error: Resampling 'custom' may not be instantiated
```
Comment (pat-s) [inline bookdown/02-basics-resampling.Rmd:348]: ?

File: bookdown/02-basics-resampling.Rmd [hunk 8]
```diff
@@ -180,3 +180,169 @@ autoplot(rr, type = "roc")
 ```
 
 All available plot types are listed on the manual page of `r ref("autoplot.ResampleResult()")`.
+
+
+### Train-Test-Validation split
+
+In many situation, we want to work with a single set of fixed train, validation and test sets. As there is confusion wrt. the exact terms used for **train**, **test** and **validation** sets, we will follow the terminology provided in this [wikipedia article](https://en.wikipedia.org/wiki/Training,_validation,_and_test_sets).
```
Comment (pat-s) [inline bookdown/02-basics-resampling.Rmd:187]:
```suggestion
In many situation, we want to work with a single set of fixed train, validation and test sets.
As there is confusion with respect to the exact terms used for **train**, **test** and **validation** sets, we will follow the terminology provided in this [wikipedia article](https://en.wikipedia.org/wiki/Training,_validation,_and_test_sets).
```

File: bookdown/02-basics-resampling.Rmd [hunk 9]
```diff
@@ -180,3 +180,169 @@ autoplot(rr, type = "roc")
...
+- **Train Set**: Used to train ML algorithms and preprocessing methods.
+
+- **Validation Set**: Used to compare and choose hyperparameters, early stopping etc.
+
+- **Test Set**: Used to obtain an un-biased estimate of the desired metric.
+
+The attentive reader has surely observed, that those concepts do not directly match to the way **mlr3** handles such data splitting procedures.
+As **mlr3** tries to stay general in order to offer a maximum of flexibility, simple splitting procedures are abstrated into **Resampling Methods** such as the ones obtained via `r ref("rsmp()")`.
+In order to tune and select hyperparameters, instead of using a **validation split**, we choose a **inner** resampling method that is used for this selection.
+This e.g. happens inside an `r ref("AutoTuner")` from `r mlr_pkg("mlr3tuning")`, where we provide an additional inner resampling method for tuning. If we now `r ref("resample()")` this `r ref("AutoTuner")`, we get the exact desired effect:
```
Comment (pat-s) [inline bookdown/02-basics-resampling.Rmd:199]:
```suggestion
This e.g. happens inside an `r ref("AutoTuner")` from `r mlr_pkg("mlr3tuning")`, where we provide an additional inner resampling method for tuning.
If we now `r ref("resample()")` this `r ref("AutoTuner")`, we get the exact desired effect:
```

**Review by pat-s — CHANGES_REQUESTED** [review]
We mainly use quotes around package names throughout the book in `library()` calls. Would be great if we could stick to this :)

File: bookdown/02-basics-resampling.Rmd (hunk 3 above)
Comment (pfistfl) [reply bookdown/02-basics-resampling.Rmd:224]: I do not understand the quotes comment

**Comment by jakob-r** [comment]
The text might be outdated because we have new API in mlr3 soon. Please update once this is implemented upstream.

**Comment by pat-s** [comment]
Stale and probably outdated, closing. Just re-open if it becomes valid again.
